nav2_amcl/src/amcl_node.cpp

```diff
@@ -914,291 +742,262 @@ AmclNode::laserReceived(sensor_msgs::msg::LaserScan::ConstSharedPtr laser_scan)
+
+  // Get the map from the map server
+  nav_msgs::msg::OccupancyGrid msg;
+  if (!map_client_.getMap(msg)) {
+    throw "Failed to get map from the map server";
   }
-*/
-  dynamic_param_client_->get_event_param("update_min_d", d_thresh_);
-  dynamic_param_client_->get_event_param("update_min_a", a_thresh_);
-  dynamic_param_client_->get_event_param("resample_interval", resample_interval_);
-  dynamic_param_client_->get_event_param("laser_min_range", laser_min_range_);
-  dynamic_param_client_->get_event_param("laser_max_range", laser_max_range_);
 
-  double save_pose_rate;
-  dynamic_param_client_->get_event_param("save_pose_rate", save_pose_rate);
-  save_pose_period = tf2::durationFromSec(1.0 / save_pose_rate);
+  RCLCPP_INFO(get_logger(), "Received a %dx%d map @ %.3f m/pix",
+    msg.info.width, msg.info.height, msg.info.resolution);
 
-  double tmp_tol;
-  dynamic_param_client_->get_event_param("transform_tolerance", tmp_tol);
-  transform_tolerance_ = tf2::durationFromSec(tmp_tol);
+  // Perform a check on the map's frame_id
+  if (msg.header.frame_id != global_frame_id_) {
+    RCLCPP_WARN(get_logger(), "Frame_id of map received:'%s' doesn't match global_frame_id:'%s'."
+      " This could cause issues with reading published topics",
+      msg.header.frame_id.c_str(), global_frame_id_.c_str());
+  }
 
-  dynamic_param_client_->get_event_param("laser_max_beams", max_beams_);
-  dynamic_param_client_->get_event_param("odom_alpha1", alpha1_);
-  dynamic_param_client_->get_event_param("odom_alpha2", alpha2_);
-  dynamic_param_client_->get_event_param("odom_alpha3", alpha3_);
-  dynamic_param_client_->get_event_param("odom_alpha4", alpha4_);
-  dynamic_param_client_->get_event_param("odom_alpha5", alpha5_);
-  dynamic_param_client_->get_event_param("laser_z_hit", z_hit_);
-  dynamic_param_client_->get_event_param("laser_z_short", z_short_);
-  dynamic_param_client_->get_event_param("laser_z_max", z_max_);
-  dynamic_param_client_->get_event_param("laser_z_rand", z_rand_);
-  dynamic_param_client_->get_event_param("laser_sigma_hit", sigma_hit_);
-  dynamic_param_client_->get_event_param("laser_lambda_short", lambda_short_);
-  dynamic_param_client_->get_event_param("laser_likelihood_max_dist",
-    laser_likelihood_max_dist_);
-  dynamic_param_client_->get_event_param("laser_model_type", sensor_model_type_);
-  dynamic_param_client_->get_event_param("robot_model_type", robot_model_type_);
-
-  dynamic_param_client_->get_event_param("min_particles", min_particles_);
-  dynamic_param_client_->get_event_param("max_particles", max_particles_);
+  // Convert to our own local data structure
+  map_ = convertMap(msg);
 
-  if (min_particles_ > max_particles_) {
-    RCLCPP_WARN(get_logger(), "You've set min_particles to be greater than max particles,"
-      " this isn't allowed so they'll be set to be equal.");
-    max_particles_ = min_particles_;
+#if NEW_UNIFORM_SAMPLING
+  // Index of free space
+  for (int i = 0; i < map_->size_x; i++) {
```

Comment:
@crdelsey Here's what I was thinking: The vector starts empty at construction. It is then populated when reading a new map (in the initMap method, which is called during on_configure). This vector is cleared, along with any other data structures during on_cleanup (the bookend match with on_configure).

Reply:
Don't we get new maps on a regular basis while in the active state? I thought AMCL has a mode where it subscribes to the map topic and gets updates regularly.

Reply:
We're missing the `use_map_topic_` functionality and subscriber to map topic in this PR.  I think this is important, so we should either re-enable this functionality here or file an issue to re-enable it later once we figure out how and if we should update maps during active state.

Reply:
@mhpanah @crdelsey The main idea is that I've avoided Activate state reconfiguration. That is, if AMCL was to receive a map over the map topic, while in the Active state, it would have a lot of recconfiguring to do, which would require using a few/several of the initialization functions. It would be counter to our goal of moving towards deterministic behavior as it would be an occasional significant hit. Instead, I'd rather move AMCL to the inactive or unconfigured state and use the code that's already there to reinitialize. Assuming that works with our expected frequency of map updates, of course. If not, we should discuss more about Activate state reconfiguring.

Reply:
Our expectation was that we'd get map updates about once per second in use cases where the map updates.

I'm not entirely sure how AMCL plays into those use cases, but it was designed to support them. I feel it is a pretty major regression in functionality if AMCL can't be used in an environment with constant map updates.

Reply:
I think it may be ok to release without AMCL supporting map updates if there is no way to get this back in, but I think we need to raise the issue to the broader team and the community and make sure everyone is OK with the tradeoff.